examples/src/main/java/org/bitcoinj/examples/ForwardingService.java

```diff
@@ -125,56 +149,41 @@ public void start() {
-            System.out.println("Transaction will be forwarded after it confirms.");
-            // Wait until it's made it into the block chain (may run immediately if it's already there).
-            //
-            // For this dummy app of course, we could just forward the unconfirmed transaction. If it were
-            // to be double spent, no harm done. Wallet.allowSpendingUnconfirmedTransactions() would have to
-            // be called in onSetupCompleted() above. But we don't do that here to demonstrate the more common
-            // case of waiting for a block.
-            waitForConfirmation(tx);
-        });
+            txFuture.complete(tx);
+        };
+        kit.wallet().addCoinsReceivedEventListener(listener);
+        return txFuture.whenComplete((tx, err) ->
+            kit.wallet().removeCoinsReceivedEventListener(listener)
+        );
     }
 
     /**
      * Wait for confirmation on a transaction.
      * @param tx the transaction we are waiting for
      */
-    void waitForConfirmation(Transaction tx) {
-        tx.getConfidence().getDepthFuture(requiredConfirmations).whenComplete((result, t) -> {
-            if (result != null) {
-                System.out.println("Confirmation received.");
-                forwardCoins(forwardingAddress);
-            } else {
-                // This kind of future can't fail, just rethrow in case something weird happens.
-                throw new RuntimeException(t);
-            }
-        });
+    CompletableFuture<TransactionConfidence> waitForConfirmation(Transaction tx) {
+        return tx.getConfidence().getDepthFuture(requiredConfirmations);
     }
 
-    private void forwardCoins(Address forwardingAddress) {
+    /**
+     * Forward the entire contents of the wallet to the forwarding address.
+     * @param transaction transaction to send
+     * @return A future for the broadcast transaction
+     */
+    CompletableFuture<Transaction> sendTransaction(Transaction transaction) {
+        SendRequest sendRequest = SendRequest.forTx(transaction);
         try {
-            // Now send the coins onwards.
-            SendRequest sendRequest = SendRequest.emptyWallet(forwardingAddress);
-            Wallet.SendResult sendResult = kit.wallet().sendCoins(sendRequest);
-            checkNotNull(sendResult);  // We should never try to send more coins than we have!
-            System.out.printf("Sending tx %s ...\n", sendResult.tx.getTxId());
-            // Register a callback that is invoked when the transaction has propagated across the network.
-            sendResult.broadcastComplete.thenAccept(transaction -> {
-                // The wallet has changed now, it'll get auto saved shortly or when the app shuts down.
-                System.out.println("Sent coins onwards!");
-            });
+            // Complete successfully when the transaction has propagated across the network.
+            return kit.wallet().sendCoins(sendRequest).broadcastComplete;
```

Comment:
Does `.sendCoins(SendRequest.forTx(transaction))` work for an already "completed" tx? If not I may have broken this. If so, sorry.

Reply:
At first, I was going to reply with "yes", but then again see below for a problem which is perhaps caused by this?

Reply:
Yes, I think that's how I broke things.

Reply:
The fix I think, is to have `sendTransaction` return a future for a `SendResult`. I was hoping to avoid exposing that type, but it has one additional important function that I neglected: it contains the completed (and technically immutable) transaction that was built by the wallet.

Reply:
That's not quite the solution either, but I'm getting closer. Stay tuned...